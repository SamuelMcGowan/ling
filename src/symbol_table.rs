use ustr::Ustr;

#[derive(Debug, Clone, Copy, serde::Serialize)]
pub(crate) struct SymbolId(usize);

#[derive(Default, Debug)]
pub(crate) struct SymbolTable {
    symbol_lookup: Vec<Symbol>,
}

impl SymbolTable {
    pub fn add(&mut self, symbol: Symbol) -> SymbolId {
        let id = self.symbol_lookup.len();
        self.symbol_lookup.push(symbol);
        SymbolId(id)
    }

    pub fn get(&self, id: SymbolId) -> Option<&Symbol> {
        self.symbol_lookup.get(id.0)
    }

    pub fn get_mut(&mut self, id: SymbolId) -> Option<&mut Symbol> {
        self.symbol_lookup.get_mut(id.0)
    }
}

#[derive(Debug)]
pub(crate) enum Symbol {
    Function { ident: Ustr },
    Var { ident: Ustr },
}

impl Symbol {
    pub fn ident(&self) -> Ustr {
        match self {
            Self::Function { ident, .. } => *ident,
            Self::Var { ident, .. } => *ident,
        }
    }
}
