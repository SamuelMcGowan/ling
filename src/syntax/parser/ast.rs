use macro_rules_attribute::{derive, derive_alias};
use ustr::Ustr;

use crate::constants::ConstIdx;

derive_alias! {
    #[derive(Node!)] = #[derive(
        Debug,
        Clone,
        serde::Serialize,
        serde::Deserialize,
    )];
}

#[derive(Node!)]
pub(crate) struct Module {
    pub items: Vec<Item>,
}

#[derive(Node!)]
pub(crate) enum Item {
    Func(Func),
}

#[derive(Node!)]
pub(crate) struct Func {
    pub ident: Ident,
    pub params: Vec<(Ident, Ty)>,
    pub ret_ty: Ty,
    pub body: Expr,
}

#[derive(Node!)]
pub(crate) enum Ty {
    Constructed { ident: Ident, params: Vec<Ty> },
    Unit,
}

#[derive(Node!)]
pub(crate) enum Stmt {
    Expr(Expr),
}

#[derive(Node!)]
pub(crate) enum Expr {
    Const(ConstIdx),
    Unit,
}

#[derive(Node!)]
pub(crate) enum Ident {
    Unresolved(Ustr),
}

impl Ident {
    pub fn unresolved(&self) -> Option<Ustr> {
        match self {
            Self::Unresolved(s) => Some(*s),
        }
    }
}
