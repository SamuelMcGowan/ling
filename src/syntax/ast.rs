use macro_rules_attribute::{derive, derive_alias};
use ustr::Ustr;

use crate::constants::ConstIdx;
use crate::symbol_table::SymbolId;

use super::parser::ParseResult;
use super::source::Span;

derive_alias! {
    #[derive(Node!)] = #[derive(
        Debug,
        Clone,
        serde::Serialize,
    )];
}

pub(crate) struct Spanned<T> {
    pub(crate) inner: T,
    pub(crate) span: Option<Span>,
}

impl<T> Spanned<ParseResult<T>> {
    pub fn transpose(self) -> ParseResult<Spanned<T>> {
        self.inner.map(|inner| Spanned {
            inner,
            span: self.span,
        })
    }
}

#[derive(Node!)]
pub(crate) struct Module {
    pub items: Vec<Item>,
}

#[derive(Node!)]
pub(crate) enum Item {
    Func(Func),
    Dummy,
}

#[derive(Node!)]
pub(crate) struct Func {
    pub ident: Ident,
    pub params: Vec<(Ident, Ty)>,
    pub ret_ty: Ty,
    pub body: Block,
}

#[derive(Node!)]
pub(crate) enum Ty {
    Constructed { ident: Ident, params: Vec<Ty> },
    Unit,
}

#[derive(Node!)]
pub(crate) struct Block {
    pub stmts: Vec<Stmt>,
    pub final_expr: Expr,
}

#[derive(Node!)]
pub(crate) enum Stmt {
    Expr(Expr),

    Loop(Block),
    WhileLoop { cond: Expr, block: Block },

    Declaration { lhs: Ident, rhs: Expr },
    Assignment { lhs: Var, rhs: Expr },

    Dummy,
}

impl Stmt {
    pub fn expect_delim(&self) -> bool {
        match self {
            Self::Expr(expr) => expr.expect_delim(),

            Self::Loop(_) | Self::WhileLoop { .. } => false,

            Self::Declaration { .. } => true,
            Self::Assignment { .. } => true,

            // the recovery function should have recovered past any semicolons
            Self::Dummy => false,
        }
    }
}

#[derive(Node!)]
pub(crate) enum Expr {
    Const(ConstIdx),

    UnaryOp {
        op: UnaryOp,
        expr: Box<Expr>,
    },

    BinOp {
        op: BinOp,
        lhs: Box<Expr>,
        rhs: Box<Expr>,
    },

    Call {
        callee: Box<Expr>,
        args: Vec<Expr>,
    },

    If {
        branches: Vec<IfBranch>,
        else_: Option<Box<Block>>,
    },

    Var(Var),

    Unit,

    Dummy,
}

impl Expr {
    pub fn expect_delim(&self) -> bool {
        !matches!(self, Self::If { .. })
    }
}

#[derive(Node!, Copy)]
pub enum UnaryOp {
    Neg,
    Not,
}

#[derive(Node!, Copy)]
pub enum BinOp {
    LogicalAnd,
    LogicalOr,

    Equal,
    NotEqual,
    Gt,
    Lt,
    GtEqual,
    LtEqual,

    Add,
    Sub,
    Mul,
    Div,
    Mod,

    Pow,

    Call,
    Access,
}

#[derive(Node!)]
pub(crate) struct IfBranch {
    pub(crate) cond: Expr,
    pub(crate) then: Block,
}

#[derive(Node!)]
pub(crate) enum Var {
    Simple(Ident),
    Field { expr: Box<Expr>, field: Ident },
}

#[derive(Node!)]
pub(crate) enum Ident {
    Unresolved(Ustr),
    Resolved(SymbolId),
}

impl Ident {
    pub fn unresolved(&self) -> Option<Ustr> {
        match self {
            Self::Unresolved(s) => Some(*s),
            Self::Resolved(_) => None,
        }
    }
}
