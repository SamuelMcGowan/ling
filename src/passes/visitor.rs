use crate::syntax::ast::*;

pub(crate) trait Visitor {
    fn visit_module(&mut self, module: &mut Module) {
        self.walk_module(module);
    }
    fn walk_module(&mut self, module: &mut Module) {
        for item in &mut module.items {
            match item {
                Item::Func(func) => self.visit_func(func),
                Item::Dummy => {}
            }
        }
    }

    fn visit_func(&mut self, func: &mut Func);

    fn visit_block(&mut self, block: &mut Block) {
        self.walk_block(block);
    }
    fn walk_block(&mut self, block: &mut Block) {
        for stmt in &mut block.stmts {
            self.visit_stmt(stmt);
        }
        self.visit_expr(&mut block.final_expr);
    }

    fn visit_stmt(&mut self, stmt: &mut Stmt) {
        self.walk_stmt(stmt);
    }
    fn walk_stmt(&mut self, stmt: &mut Stmt) {
        match stmt {
            Stmt::Expr(expr) => self.visit_expr(expr),

            Stmt::Loop(block) => self.visit_loop(block),
            Stmt::WhileLoop { cond, block } => self.visit_while_loop(cond, block),

            Stmt::Declaration { lhs, rhs } => self.visit_declaration(lhs, rhs),
            Stmt::Assignment { lhs, rhs } => self.visit_assignment(lhs, rhs),

            Stmt::Dummy => {}
        }
    }

    fn visit_expr(&mut self, expr: &mut Expr) {
        self.walk_expr(expr);
    }
    fn walk_expr(&mut self, expr: &mut Expr) {
        match expr {
            Expr::Const(_) => {}
            Expr::UnaryOp { op: _, expr } => self.visit_expr(expr),
            Expr::BinOp { op: _, lhs, rhs } => {
                self.visit_expr(lhs);
                self.visit_expr(rhs);
            }
            Expr::Call { callee, args } => {
                self.visit_expr(callee);
                for arg in args {
                    self.visit_expr(arg);
                }
            }
            Expr::If { branches, else_ } => {
                for branch in branches {
                    self.visit_expr(&mut branch.cond);
                    self.visit_block(&mut branch.then);
                }
                if let Some(block) = else_ {
                    self.visit_block(block);
                }
            }
            Expr::Var(var) => self.visit_var(var),

            Expr::Unit => {}

            Expr::Dummy => {}
        }
    }

    fn visit_declaration(&mut self, lhs: &mut Ident, rhs: &mut Expr);

    fn visit_assignment(&mut self, lhs: &mut Var, rhs: &mut Expr) {
        self.visit_var(lhs);
        self.visit_expr(rhs);
    }

    fn visit_var(&mut self, var: &mut Var);

    fn visit_loop(&mut self, block: &mut Block) {
        self.visit_block(block)
    }

    fn visit_while_loop(&mut self, cond: &mut Expr, block: &mut Block) {
        self.visit_expr(cond);
        self.visit_block(block);
    }
}
