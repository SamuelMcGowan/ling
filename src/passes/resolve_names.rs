use ustr::Ustr;

use super::visitor::Visitor;
use crate::ast::*;
use crate::symbol_table::{Symbol, SymbolId, SymbolKind, SymbolTable};

#[derive(Debug)]
pub(crate) enum SymbolError {
    SymbolNotFound(Ustr),
    GlobalShadowed(Ustr),
    WrongKind(Ustr),
}

pub(crate) struct Resolver {
    table: SymbolTable,

    builtins: Vec<SymbolEntry>,
    globals: Vec<SymbolEntry>,
    locals: Vec<SymbolEntry>,

    scopes: Vec<usize>,

    errors: Vec<SymbolError>,
}

impl Resolver {
    pub fn new() -> Self {
        let mut r = Self {
            table: SymbolTable::default(),

            builtins: vec![],
            globals: vec![],
            locals: vec![],

            scopes: vec![],

            errors: vec![],
        };

        r.declare_builtin("int", SymbolKind::TyStruct);
        r.declare_builtin("uint", SymbolKind::TyStruct);

        r
    }

    pub fn visit(ast: &mut Module) -> (SymbolTable, Vec<SymbolError>) {
        let mut resolver = Self::new();
        resolver.visit_module(ast);
        (resolver.table, resolver.errors)
    }

    fn declare(&mut self, ident: Ustr, kind: SymbolKind) -> SymbolEntry {
        let is_value = kind.is_value();

        let symbol_id = self.table.add(Symbol { ident, kind });

        SymbolEntry {
            ident,
            symbol_id,
            is_value,
        }
    }

    fn declare_builtin(&mut self, ident: impl Into<Ustr>, kind: SymbolKind) {
        let entry = self.declare(ident.into(), kind);
        self.builtins.push(entry);
    }

    #[must_use]
    fn declare_global(&mut self, ident: Ustr, kind: SymbolKind) -> Ident {
        // check a global isn't already declared with this name.
        if self.globals.iter().any(|entry| entry.ident == ident) {
            self.errors.push(SymbolError::GlobalShadowed(ident));
            return Ident::Unresolved(ident);
        }

        let entry = self.declare(ident, kind);
        self.globals.push(entry);

        Ident::Resolved(entry.symbol_id)
    }

    #[must_use]
    fn declare_local(&mut self, ident: Ustr, kind: SymbolKind) -> Ident {
        let entry = self.declare(ident, kind);
        self.locals.push(entry);

        Ident::Resolved(entry.symbol_id)
    }

    #[must_use]
    fn resolve(&mut self, ident: Ustr, value: bool) -> Ident {
        let Some(entry) = self.resolve_entry(ident) else {
            self.errors.push(SymbolError::SymbolNotFound(ident));
            return Ident::Unresolved(ident);
        };

        if entry.is_value != value {
            self.errors.push(SymbolError::WrongKind(ident));
            return Ident::Unresolved(ident);
        }

        Ident::Resolved(entry.symbol_id)
    }

    fn resolve_entry(&self, ident: Ustr) -> Option<&SymbolEntry> {
        fn find_entry(ident: Ustr, entries: &[SymbolEntry]) -> Option<&SymbolEntry> {
            entries.iter().rev().find(|entry| entry.ident == ident)
        }

        find_entry(ident, &self.locals)
            .or_else(|| find_entry(ident, &self.globals))
            .or_else(|| find_entry(ident, &self.builtins))
    }

    fn push_scope(&mut self) {
        self.scopes.push(self.locals.len());
    }

    fn pop_scope(&mut self) {
        let start_len = self.scopes.pop().unwrap();
        self.locals.truncate(start_len);
    }
}

impl Resolver {
    fn forward_declare_item(&mut self, item: &mut Item) {
        match item {
            Item::Func(func) => {
                func.ident =
                    self.declare_global(func.ident.unresolved().unwrap(), SymbolKind::Function);
            }
            Item::Struct(strukt) => {
                strukt.ident =
                    self.declare_global(strukt.ident.unresolved().unwrap(), SymbolKind::TyStruct);
            }
            Item::Enum(eenum) => {
                eenum.ident =
                    self.declare_global(eenum.ident.unresolved().unwrap(), SymbolKind::TyEnum)
            }
            Item::Dummy => {}
        }
    }

    fn declare_ty_params(&mut self, ty_params: &mut [Ident]) {
        for ident in ty_params {
            *ident = self.declare_local(ident.unresolved().unwrap(), SymbolKind::TyParam);
        }
    }
}

impl Visitor for Resolver {
    fn walk_module(&mut self, module: &mut Module) {
        for item in &mut module.items {
            self.forward_declare_item(item);
        }

        for item in &mut module.items {
            self.visit_item(item);
        }
    }

    fn visit_func(&mut self, func: &mut Func) {
        // already declared

        self.push_scope();

        self.declare_ty_params(&mut func.ty_params);

        for (ident, ty) in &mut func.params {
            *ident = self.declare_local(ident.unresolved().unwrap(), SymbolKind::Var);
            self.visit_ty(ty);
        }
        self.visit_ty(&mut func.ret_ty);

        self.visit_block(&mut func.body);

        self.pop_scope();
    }

    fn visit_struct(&mut self, strukt: &mut Struct) {
        // already declared

        self.push_scope();

        self.declare_ty_params(&mut strukt.ty_params);

        for field in &mut strukt.fields {
            self.visit_struct_field(field);
        }

        self.pop_scope();
    }

    fn visit_struct_field(&mut self, field: &mut StructField) {
        self.visit_ty(&mut field.ty);
    }

    fn visit_enum(&mut self, eenum: &mut Enum) {
        // already declared

        self.push_scope();

        self.declare_ty_params(&mut eenum.ty_params);

        for variant in &mut eenum.variants {
            self.visit_enum_variant(variant);
        }

        self.pop_scope();
    }

    fn visit_enum_variant(&mut self, variant: &mut EnumVariant) {
        self.visit_enum_variant_kind(&mut variant.kind);
    }

    fn visit_block(&mut self, block: &mut Block) {
        self.push_scope();
        self.walk_block(block);
        self.pop_scope();
    }

    fn visit_var(&mut self, var: &mut Var) {
        match var {
            Var::Simple(ident) => *ident = self.resolve(ident.unresolved().unwrap(), true),
            Var::Field { expr: _, field: _ } => {
                // TODO
            }
        }
    }

    fn visit_declaration(&mut self, lhs: &mut Ident, rhs: &mut Expr) {
        // visit expression first so that identifier can't appear in expression.
        self.visit_expr(rhs);
        *lhs = self.declare_local(lhs.unresolved().unwrap(), SymbolKind::Var);
    }

    fn visit_ty(&mut self, ty: &mut Ty) {
        match ty {
            Ty::Tuple(fields) => {
                for field in fields {
                    self.visit_ty(field)
                }
            }
            Ty::Constructed { ident, params } => {
                *ident = self.resolve(ident.unresolved().unwrap(), false);
                for param in params {
                    self.visit_ty(param);
                }
            }
        }
    }
}

#[derive(Debug, Clone, Copy)]
struct SymbolEntry {
    ident: Ustr,
    symbol_id: SymbolId,
    is_value: bool,
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::{Resolver, SymbolError};
    use crate::ast::Module;
    use crate::parser::test_parse;
    use crate::symbol_table::SymbolTable;

    fn test_resolve(source: &str) -> (Module, SymbolTable, Vec<SymbolError>) {
        let (mut ast, mismatched_brackets, parse_errors) = test_parse(source, |p| p.parse_module());

        if !mismatched_brackets.is_empty() {
            panic!("mismatched brackets: {mismatched_brackets:?}");
        }

        if !parse_errors.is_empty() {
            panic!("parse errors: {parse_errors:?}");
        }

        let (table, errors) = Resolver::visit(&mut ast);

        (ast, table, errors)
    }

    #[test]
    fn simple() {
        assert_debug_snapshot!(test_resolve("func foo() {}"));
    }

    #[test]
    fn shadowed_global() {
        assert_debug_snapshot!(test_resolve("func foo() { foo() } func foo() {}"));
    }

    #[test]
    fn shadowed_local() {
        assert_debug_snapshot!(test_resolve("func foo() { let a = 12; let a = 12; }"));
    }

    #[test]
    fn shadowed_global_shadowed_local() {
        assert_debug_snapshot!(test_resolve(
            "func foo() { let a = 12; let a = 12; } func foo() {}"
        ));
    }

    #[test]
    fn local_shadowing_global() {
        assert_debug_snapshot!(test_resolve("func foo() { let foo = 12; let a = foo; }"))
    }

    #[test]
    fn global_shadowing_builtin() {
        assert_debug_snapshot!(test_resolve("func uint() { uint() }"))
    }

    #[test]
    fn forward_declaration() {
        assert_debug_snapshot!(test_resolve("func a() { a(); b(); } func b() {}"))
    }

    #[test]
    fn argument() {
        assert_debug_snapshot!(test_resolve("func foo(a: uint) { a } func a() {}"));
    }

    #[test]
    fn argument_shadowed() {
        assert_debug_snapshot!(test_resolve("func foo(a: uint) { let a = 12; }"))
    }

    #[test]
    fn func_generic() {
        assert_debug_snapshot!(test_resolve("func foo[A, B]() {}"));
    }
    #[test]
    fn strukt() {
        assert_debug_snapshot!(test_resolve("data Person { name: string, age: uint }"));
    }

    #[test]
    fn eenum() {
        assert_debug_snapshot!(test_resolve("enum Result[T, E] { Ok(T), Err(E), }"));
    }

    #[test]
    fn tuple() {
        assert_debug_snapshot!(test_resolve("func a() -> (uint, uint) {}"));
    }

    #[test]
    fn assign_to_generic() {
        assert_debug_snapshot!(test_resolve("func foo[A]() { A = 12; }"));
    }

    #[test]
    fn resolution() {
        assert_debug_snapshot!(test_resolve("func foo(a: uint) { let b = a; let c = b; }"));
    }

    #[test]
    fn scopes() {
        assert_debug_snapshot!(test_resolve("func foo() { loop { let a = 12; } a;}"));
    }

    #[test]
    fn unknown_var() {
        assert_debug_snapshot!(test_resolve("func foo() { bloop() }"))
    }
}
