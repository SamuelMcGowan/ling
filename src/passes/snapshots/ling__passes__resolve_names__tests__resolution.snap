---
source: src/passes/resolve_names.rs
expression: "test_resolve(\"func foo(a: uint) { let b = a; let c = b; }\")"
---
(
    Module {
        items: [
            Dummy,
        ],
    },
    SymbolTable {
        symbol_lookup: [
            Symbol {
                ident: u!("int"),
                kind: TyStruct(
                    Builtin,
                ),
            },
            Symbol {
                ident: u!("uint"),
                kind: TyStruct(
                    Builtin,
                ),
            },
            Symbol {
                ident: u!("foo"),
                kind: Function(
                    Ast(
                        Func {
                            ident: Resolved(
                                SymbolId(
                                    2,
                                ),
                            ),
                            ty_params: [],
                            params: [
                                (
                                    Resolved(
                                        SymbolId(
                                            3,
                                        ),
                                    ),
                                    Constructed {
                                        ident: Resolved(
                                            SymbolId(
                                                1,
                                            ),
                                        ),
                                        params: [],
                                    },
                                ),
                            ],
                            ret_ty: Tuple(
                                [],
                            ),
                            body: Block {
                                stmts: [
                                    Declaration {
                                        lhs: Resolved(
                                            SymbolId(
                                                4,
                                            ),
                                        ),
                                        rhs: Var(
                                            Simple(
                                                Resolved(
                                                    SymbolId(
                                                        3,
                                                    ),
                                                ),
                                            ),
                                        ),
                                    },
                                    Declaration {
                                        lhs: Resolved(
                                            SymbolId(
                                                5,
                                            ),
                                        ),
                                        rhs: Var(
                                            Simple(
                                                Resolved(
                                                    SymbolId(
                                                        4,
                                                    ),
                                                ),
                                            ),
                                        ),
                                    },
                                ],
                                final_expr: Unit,
                            },
                        },
                    ),
                ),
            },
            Symbol {
                ident: u!("a"),
                kind: Var,
            },
            Symbol {
                ident: u!("b"),
                kind: Var,
            },
            Symbol {
                ident: u!("c"),
                kind: Var,
            },
        ],
    },
    [],
)
